File: ChatGPT/Level2_Task_GPT/Level4_Prompt_GPT/VocabularyTrainer.cpp
#include <iostream>
#include <unordered_map>
#include <cstdlib>  // For rand() and srand()
#include <ctime>    // For time()
#include <algorithm> // For case-insensitive comparison

// Function to add a new vocabulary to the map
void addVocabulary(std::unordered_map<std::string, std::string>& vocabularyMap) {
    // Prompt user to insert vocabulary
    std::cout << "Enter a vocabulary (word or sentence): ";
    std::string vocabulary;
    std::getline(std::cin, vocabulary);

    // Prompt user to insert the meaning of the vocabulary
    std::cout << "Enter the meaning of the vocabulary: ";
    std::string meaning;
    std::getline(std::cin, meaning);

    // Store the vocabulary and its meaning in the map
    vocabularyMap[vocabulary] = meaning;

    std::cout << "\nVocabulary added successfully!\n\n";
}

// Function to conduct a learning session with stored vocabularies
void learnVocabularies(std::unordered_map<std::string, std::string>& vocabularyMap) {
    if (vocabularyMap.empty()) {
        std::cout << "No vocabularies available for learning. Add some vocabularies first.\n\n";
        return;
    }

    // Seed for randomization
    std::srand(std::time(0));

    // Iterate through all stored vocabularies
    for (const auto& pair : vocabularyMap) {
        // Display the vocabulary
        std::cout << "Vocabulary: " << pair.first << "\n";

        // Prompt user to input the meaning
        std::cout << "Enter the meaning of the vocabulary: ";
        std::string userMeaning;
        std::getline(std::cin, userMeaning);

        // Convert both the stored meaning and user input to lowercase for case-insensitive comparison
        std::string storedMeaningLower = pair.second;
        std::transform(storedMeaningLower.begin(), storedMeaningLower.end(), storedMeaningLower.begin(), ::tolower);
        std::transform(userMeaning.begin(), userMeaning.end(), userMeaning.begin(), ::tolower);

        // Check if the user's input matches the stored meaning
        if (userMeaning == storedMeaningLower) {
            std::cout << "Correct!\n\n";
        } else {
            std::cout << "Wrong! The correct meaning is: " << pair.second << "\n\n";
        }
    }

    std::cout << "Learning session complete!\n\n";
}

// Main function
int main() {
    std::unordered_map<std::string, std::string> vocabularyMap;

    int choice;
    bool learned = false;

    do {
        // Display menu
        std::cout << "Menu:\n";
        std::cout << "1. Add new vocabulary\n";
        std::cout << "2. Learn vocabularies\n";
        std::cout << "3. Exit\n";

        if (!learned) {
            std::cout << "Note: Complete learning session (option 2) before choosing other options.\n";
        }

        std::cout << "Enter your choice (1-3): ";

        // Get user choice
        std::cin >> choice;
        std::cin.ignore();  // Ignore the newline character in the input buffer

        // Perform actions based on user choice
        switch (choice) {
            case 1:
                addVocabulary(vocabularyMap);
                break;
            case 2:
                learnVocabularies(vocabularyMap);
                learned = true;
                break;
            case 3:
                std::cout << "Exiting the Vocabulary Trainer. Goodbye!\n";
                break;
            default:
                std::cout << "Invalid choice. Please enter a number between 1 and 3.\n\n";
        }

    } while (choice != 3);

    return 0;
}

File: ChatGPT/Level2_Task_GPT/Level1_Prompt_GPT/vocabulary_trainer.py
class VocabularyTrainer:
    def __init__(self):
        self.vocabulary = {}

    def add_word(self, word, definition):
        self.vocabulary[word] = definition

    def learn_words(self):
        if not self.vocabulary:
            print("No words in the vocabulary. Add words first.")
            return

        print("Welcome to the Vocabulary Trainer!")
        for word, definition in self.vocabulary.items():
            user_input = input(f"What does '{word}' mean? ")
            if user_input.lower() == definition.lower():
                print("Correct!")
            else:
                print(f"Wrong! The correct definition is: {definition}")

    def display_vocabulary(self):
        if not self.vocabulary:
            print("No words in the vocabulary.")
            return

        print("Vocabulary:")
        for word, definition in self.vocabulary.items():
            print(f"{word}: {definition}")


def main():
    trainer = VocabularyTrainer()

    while True:
        print("\n1. Add Word\n2. Learn Vocabulary\n3. Display Vocabulary\n4. Exit")
        choice = input("Enter your choice (1/2/3/4): ")

        if choice == "1":
            word = input("Enter the word: ")
            definition = input(f"Enter the definition of '{word}': ")
            trainer.add_word(word, definition)
            print(f"'{word}' added to the vocabulary.")

        elif choice == "2":
            trainer.learn_words()

        elif choice == "3":
            trainer.display_vocabulary()

        elif choice == "4":
            print("Exiting the Vocabulary Trainer. Goodbye!")
            break

        else:
            print("Invalid choice. Please enter a valid option.")


if __name__ == "__main__":
    main()
